Maurits big cleanup

mauritsvanrees opened the pull request:
Earlier today, Zope removed lots of version pins for packages that core Zope should not care about. We should do the same for Plone. This PR does the following:

- Cleanup of old no longer used settings and files.
- Completely restructure `versions.cfg`, splitting non-core versions out into `versions-extra.cfg`
- To keep a bit of structure in the still long list in `versions.cfg`, I split them into four sections based on GitHub organisations: plone, collective, zope/zodb/z3c/repoze, and other.
- I added `bare.cfg` plus an extra GHA yaml file to check the the pins in `versions.cfg` are enough to create a Plone instance plus a `bin/test` script.

mauritsvanrees committed:
Removed cmf.cfg and plone.versioncheck-dev.cfg from experimental.

mauritsvanrees committed:
Removed definitions of template/config/scripts-directory.

These are not used in current coredev.

mauritsvanrees committed:
Do not add Plone to the eggs in plipbase.

This was a change from 2014 which was somehow needed for Archetypes.

mauritsvanrees committed:
Use less allow-hosts and remove find-links.

mauritsvanrees committed:
Remove several version pins, move non-core ones to versions-extra.cfg.

versions.cfg should now only have pins for packages used when creating a Plone instance and `bin/test`.
I have split them into four groups to have a bit of structure.
The four groups are based on the GitHub organisations that the packages belong to:

1. Plone
2. Collective
3. Zope, ZODB, z3c, repoze
4. other

Fine to revise if this becomes hard to understand and maintain.

mauritsvanrees committed:
Pin fancycompleter

mauritsvanrees committed:
Pin pyrepl

jensens commented:
Overall big love for this! But I would not mix up core + ecosystem packages (like mosaic). I tend to add a versions-ecosystem file unless those are not officially core. What do you think?

mauritsvanrees committed:
fixed plone.app.contentlisting pin

mauritsvanrees commented:
> Overall big love for this! But I would not mix up core + ecosystem packages (like mosaic). I tend to add a versions-ecosystem file unless those are not officially core. What do you think?

That can work.
But in this PR, there are not yet any ecosystem packages. All versions in `versions.cfg` are used while creating `bin/instance` and `bin/test`. For example I did not expect `plone.tiles` in here, but `plone.restapi` pulls it in as test dependency.

mauritsvanrees committed:
Added more missing pins when using older Pythons.

Python 3.9 does not need these, 3.7 does.

mauritsvanrees committed:
Add versions-ecosystem.cfg

mauritsvanrees committed:
prefer-final = false

mauritsvanrees committed:
bare.yml: generate locale

Needed for CMFPlone testUnicodeSplitter test on GitHub Actions:

```
Error in test testNormalizeLatin1 (Products.CMFPlone.tests.testUnicodeSplitter.TestCaseNormalizer)
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/home/runner/work/buildout.coredev/buildout.coredev/src/Products.CMFPlone/Products/CMFPlone/tests/testUnicodeSplitter.py", line 144, in testNormalizeLatin1
    saved = _setlocale(*LATIN1)
  File "/home/runner/work/buildout.coredev/buildout.coredev/src/Products.CMFPlone/Products/CMFPlone/tests/testUnicodeSplitter.py", line 37, in _setlocale
    "available on your system: %s" % str(LATIN1)
ValueError: Unsupported locale. These tests need at least one of the following locales available on your system: ('en_US.ISO-8859-1', 'en_US.ISO8859-15', 'en_GB.ISO8859-15', 'de_DE@euro', 'fr_FR@euro', 'nl_NL@euro')
```

mauritsvanrees committed:
bare.yml: actually use bare.cfg as buildout config.

mauritsvanrees committed:
Add ecosystem.yml to test ecosystem.cfg.

mauritsvanrees committed:
ecosystem: checkout datagridfield and mosaic

mauritsvanrees committed:
release tox: include versions-ecosystem.cfg

mauritsvanrees committed:
Cleanup ecosystem.cfg

mauritsvanrees committed:
release tox: get constraints from ecosystem.cfg.

This includes all other version config files.

mauritsvanrees committed:
Removed tox.ini in top level directory.

This was far outdated, so no one could have been using this on Plone 6.
It *could* be nice if this allows us to have the source checkouts only once and still test on different Python versions.

mauritsvanrees commented:
I have added `versions-ecosystem.cfg` and updated  `ecosystem.cfg` to use this.

mauritsvanrees committed:
bare.yml: fix setting locale

mauritsvanrees committed:
GitHub Actions: use buildout:git-clone-depth=1

mauritsvanrees committed:
Update GHA to use latest actions, and cache pip.

Use defaults for py-actions/py-dependency-install

mauritsvanrees committed:
GHA: only take hash of versions*.cfg for cache key.

mauritsvanrees commented:
I have improved the GitHub Actions. Should cache better the next time.
Now Jenkins:

@jenkins-plone-org please run jobs

ale-rt reviewed (approved):
Wonderful job :)

Some consideration you might ignore or that can be handles another time:
- should core.cfg extend bare.cfg?
- should we also have a sources-extra.cfg?

mauritsvanrees committed:
Let core.cfg extend bare.cfg.

mauritsvanrees committed:
Reorder sources.cfg

Moved all non-eggs together.
Alphabetically sorted the rest, without keeping the Products separate.
Removed the checkouts that are already defined in the Zope sources.cfg.
Note that Zope itself is not actually included in the Zope sources.cfg, so we add it here. :-)

mauritsvanrees commented:
> * should core.cfg extend bare.cfg?

That is an option. Okay, I have changed it.

> * should we also have a sources-extra.cfg?

I think it is clearer to have all sources in one file, so developers do not have to hunt for where a source is defined. I don't expect that we will define many extra sources.
I have reordered `sources.cfg` a bit now, putting the non-eggs apart, and removing sources that are already define in the Zope sources.

mauritsvanrees committed:
Try to improve caching of GitHub Actions.

I saw caches getting restored, and then buildout still taking more than 15 minutes.
I expect this is because we restore a cache with for example Python 3.7 eggs in a Python 3.9 job.

mauritsvanrees commented:
Meanwhile trying to improve the caching of GHA further. Previous run was all over the place for how long a run took, some around 2 minutes, some 15, some almost 30:

![Screenshot 2022-04-14 at 10 31 24](https://user-images.githubusercontent.com/210587/163346746-1f7a9912-e066-4d05-99f8-8c11fcffa622.png)

I am trying this now in all yaml files:

```
- name: Cache eggs
  uses: actions/cache@v3
  with:
    path: |
      eggs
    key: eggs-${{ matrix.python-version }}-${{ runner.os }}-${{ hashFiles('versions*.cfg') }}
    restore-keys: |
      eggs-${{ matrix.python-version }}-${{ runner.os }}-
      eggs-${{ matrix.python-version }}-
```

mauritsvanrees committed:
vscode: do not use the PyLance server, because it is no longer supported.

I have switched to the new default languge server a long time ago,
and ever since, VSCode kept showing a popup to install PyLance, apparently because the recipe settings ask for it.

mauritsvanrees commented:
Another run after my cache changes shows around 15 minutes for each, as none of them now find a cache:

![Screenshot 2022-04-14 at 10 51 27](https://user-images.githubusercontent.com/210587/163350592-84688fdf-9a17-483f-8997-2a00ebbe81e3.png)

I did another small change, configuring the vscode recipe, and in the next run all should be a lot faster.

mauritsvanrees commented:
That is better. Longest was 6 minutes:

![Screenshot 2022-04-14 at 11 02 02](https://user-images.githubusercontent.com/210587/163351834-a7ef16b1-a048-44b5-b443-2cee3914af24.png)

Cache extraction works:

```
Cache restored from key:
eggs-3.7-Linux-4c6074f594712aef4a7d92010ae65d7d4dd84e7c7c7f268628e46071eaa03b65
```

@jenkins-plone-org please run jobs

ale-rt commented:
Merge this at will :)

mauritsvanrees commented:
Merged. Thanks for the reviews.

gforcada commented on sources.cfg:
```diff
@@ -164,26 +155,30 @@ Products.ExternalEditor             = git ${remotes:zope}/Products.ExternalEdito
 Products.validation                 = git ${remotes:plone}/Products.validation.git pushurl=${remotes:plone_push}/Products.validation.git branch=master
-Products.ZCatalog                   = git ${remotes:zope}/Products.ZCatalog.git pushurl=${remotes:zope_push}/Products.ZCatalog.git branch=master
 Products.ZopeVersionControl         = git ${remotes:zope}/Products.ZopeVersionControl.git pushurl=${remotes:zope_push}/Products.ZopeVersionControl.git branch=master
+repoze.xmliter                      = git https://github.com/repoze/repoze.xmliter.git pushurl=git@github.com:repoze/repoze.xmliter.git branch=master
```
We could use here the `${remotes:zope}` :)

mauritsvanrees replied:
That would then need to be `remotes:repoze`, which does not exist yet. And since it is the only package we use from repoze, let's not add the extra indirection.

gforcada replied:
oh my 🤦🏾 sorry, I was not reading properly 😓 sorry for the noise!

gforcada commented on versions-ecosystem.cfg:
```diff
@@ -0,0 +1,22 @@
+# Read the documentation of the individual packages, or ask the maintainers.
+[versions]
+# collective.z3cform.datagridfield = 2.0.2
+# collective.z3cform.datetimewidget = 1.2.9
```
you kept these pins commented as they are not needed, but might be worth keeping them around as we might want to uncomment them eventually? 🤔

mauritsvanrees replied:
I kept them because it would be good to have version pins for these, as they were in the 5.2 pins as well. But there are no released versions of these two packages yet that work on current Plone 6.
Some work is done here:

- https://github.com/collective/collective.z3cform.datetimewidget/pull/12
- https://github.com/collective/collective.z3cform.datagridfield/pull/120

gforcada commented:
Sorry to be late on the party 😅 I was on vacations with my kids 🎉 

I could only mentions these two minor issues/questions/fixes, great job @mauritsvanrees ! 💯
